src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeReader.java

```diff
@@ -0,0 +1,455 @@
+
+            int nodeID = index.getNodeID();
+
+            if (postingsIndex.exists(nodeID) && fixedBitSet.cardinality() > 0)
+            {
+                long pointer = postingsIndex.getPostingsFilePointer(nodeID);
+                postingLists.add(initFilteringPostingReader(pointer, fixedBitSet));
+            }
+        }
+
+        void visitNode(PriorityQueue<PeekablePostingList> postingLists, byte[] cellMinPacked, byte[] cellMaxPacked) throws IOException
+        {
+            byte[] splitPackedValue = index.getSplitPackedValue();
+            BytesRef splitDimValue = index.getSplitDimValue();
+            assert splitDimValue.length == bytesPerValue;
+
+            // make sure cellMin <= splitValue <= cellMax:
+            assert FutureArrays.compareUnsigned(cellMinPacked, 0, bytesPerValue, splitDimValue.bytes, splitDimValue.offset, splitDimValue.offset + bytesPerValue) <= 0 : "bytesPerValue=" + bytesPerValue;
+            assert FutureArrays.compareUnsigned(cellMaxPacked, 0, bytesPerValue, splitDimValue.bytes, splitDimValue.offset, splitDimValue.offset + bytesPerValue) >= 0 : "bytesPerValue=" + bytesPerValue;
+
+            // Recurse on left subtree:
+            System.arraycopy(cellMaxPacked, 0, splitPackedValue, 0, packedBytesLength);
+            System.arraycopy(splitDimValue.bytes, splitDimValue.offset, splitPackedValue, 0, bytesPerValue);
+
+            index.pushLeft();
+            collectPostingLists(postingLists, cellMinPacked, splitPackedValue);
+            index.pop();
+
+            // Restore the split dim value since it may have been overwritten while recursing:
+            System.arraycopy(splitPackedValue, 0, splitDimValue.bytes, splitDimValue.offset, bytesPerValue);
+            // Recurse on right subtree:
+            System.arraycopy(cellMinPacked, 0, splitPackedValue, 0, packedBytesLength);
+            System.arraycopy(splitDimValue.bytes, splitDimValue.offset, splitPackedValue, 0, bytesPerValue);
+            index.pushRight();
+            collectPostingLists(postingLists, splitPackedValue, cellMaxPacked);
+            index.pop();
+        }
+
+        @SuppressWarnings({"resource", "RedundantSuppression"})
+        private PeekablePostingList initFilteringPostingReader(long offset, FixedBitSet filter) throws IOException
+        {
+            final PostingsReader.BlocksSummary summary = new PostingsReader.BlocksSummary(postingsSummaryInput, offset);
+            PostingsReader postingsReader = new PostingsReader(postingsInput, summary, listener.postingListEventListener());
+            return PeekablePostingList.makePeekable(new FilteringPostingList(filter, postingsReader));
+        }
+
+        private FixedBitSet visitDocValues(IndexInput in, int count, IntersectVisitor visitor, short[] origIndex) throws IOException
+        {
+            int commonPrefixLength = readCommonPrefixLength(in);
+            int compressedDimension = readCompressedDimension(in);
```

Comment:
We have the common prefix length and the logical bytes per dimension already. Why do we need the "compressed dimension" at all in the on-disk format?

Reply:
Ok, this is nothing to do with dimensions. It indicates whether all the values in a leaf are the same or not. I have changed its name and implementation to be `boolean singleValue`. I have to admit that this one took me a long time to figure out what it was trying to do.